packages/nimble-components/build/generate-workers/source/matrix-renderer.ts

```diff
@@ -8,16 +9,179 @@ import { expose } from 'comlink';
  * This setup is used in the wafer-map component to perform heavy computational duties
  */
 export class MatrixRenderer {
-    public dieMatrix: Uint8Array = Uint8Array.from([]);
+    public columnIndexes = Int32Array.from([]);
```

Comment:
There is a lot of state and operations being added to this file. Considerable work was done to break-up the logic for the main wafer-map code into different parts (data-manager, prerendering, rendering). Is there a clear idea of how the logic in here will be organized? We won't end up with one giant class will we? @Razvan1928 @munteannatan

Reply:
Considering the changes that we have in plan to push for the wafer map despite the ones inside this pr (adding text on dies, coloring the dies and dies highlighting) I expect the file to reach 300-350 lines, is it too much or should we try to split it into multiple files?

Reply:
... it's not about number of lines, it's about separation of respponsibilities and consistency.

If "MatrixRenderer" is analogous in responsibilities to ["RenderingModule"](https://github.com/ni/nimble/blob/main/packages/nimble-components/src/wafer-map/modules/rendering.ts) then right now if I compare the two they look very different. Rendering Module does not have all state dumped in one file, instead element specific state is on a class representing the Elements state (WaferMap), there is a reference to dataManager which references computations and preerendering, etc.

I'd like to understand at a high-level why this Rendering class would be so different and not try to orgnaize state and logic similarly.

Reply:
I'm trying to find a way to get rid of that much state, but I don't yet have a solution, I'm gonna explain you the situation and maybe find a solution together. (or leave it as it as, we'll see) Most of the properties of `matrix-renderer.ts` are the same as the original wafermap, so passing a wafer through the constructor would get rid of a lot of state, but I cannot pass a WaferMap object to `matrix-renderer.ts` as the directory `nimble-components\build` is build before `nimble-components\src` and cannot reference any class from there. Creating a new `WaferMap` type inside `nimble-components\build` directory doesn't sound like a good idea. `nimble-components\build\generate-workers` should be built before because rollup packs up the `dist` folder of `generate-workers` and `index.ts` moves it to `nimble-components\src\wafer-map\workers\matrix-renderer.ts`, file without which by the time `src/wafer-map` is build will cause an error, as `src/wafer-map/modules/create-matrix-renderer.ts` references `nimble-components\src\wafer-map\workers\matrix-renderer.ts`. Regarding the separation of concerns, we somehow didn't think about doing it for `matrix-renderer.ts` , I think we can split the file similarly to how wafer-map is organized rendering/prerendering/data-manager etc, will try to do a prototype to see how it works. Unfortunately @munteannatan is OOO until Monday and can't give advise.

Reply:
Discussed in person and clarified that the goal isn't necessarily to share the exact same code but to follow the same patterns. It does not make sense to try and leverage the Wafer Map Element code in the worker since there is no DOM in the worker, etc. Instead the goal is to have similar organizational constructs by splitting the state and logic into different parts.

This stackblitz shows a high-level example of splitting up the state: https://stackblitz.com/edit/comlink-split-data?file=src%2Fmain.ts,src%2Fouter-state.ts,src%2Finner-state.ts
- The example has an outer object that creates an inner state object (there could be multiple such as data-manager, prerendering, rendering, etc)
- Then the inner state object is proxied to the main thread. The main thread is able to set state on the inner state objects and invoke methods on the outer state object which uses the inner state synchronously in the same thread.

Reply:
As was discussed yesterday with @munteannatan this will be handled in a different PR